text/0757-default-modifier-manager.md

```diff
@@ -0,0 +1,276 @@
+[rfc-779]:https://github.com/emberjs/rfcs/pull/779
+
+This pairs nicely with the [First-Class Component Templates RFC](rfc-779);
+where modifiers may be defined in the same file as a template-only component.
+
+<details><summary>Example</summary>
+
+<!--
+  If you're reading this RFC non-rendered, ignore the "jsx" language tag.
+  "jsx" is incorrect, and isn't perfect, but "gjs" syntax highlighting
+  hasn't been added to GitHub's highlighting yet.
+-->
+
+This could be an entire component
+
+```jsx
+const draggable = element => {
+  element.addEventListener( /*... */ );
+
+  return () => element.removeEventListener( /* ... */ );
+};
+
+<template>
+  <div {{draggable}}>text content A</div>
+  <div {{draggable}}>text content B</div>
+  <div {{draggable}}>text content C</div>
+  <div {{draggable}}>text content D</div>
+</template>
+```
+_Note: `<template>` is experimental syntax, and should not be be the focus of this example_
+
+
+</details>
+
+
+## Detailed design
+
+_A Default Manager is not something that can be chosen by the user, but is baked in to the framework
+as a default so that a user doesn't have to build something to use a non-framework-specific variant
+of the three constructs: Helpers, Modifiers, and Components._
+
+The desired usage of a plain function in a template should be:
+ - convenient
+ - reduce boilerplate
+ - be easily portable to JS for developers' mental model of how template and JS interact.
+
+Which results in:
+ - default to positional parameters
+ - all named arguments are grouped into an "options object" as the last parameter.
+   this happens to align with the _syntax_ of modifier invocation where named arguments may not appear
+   before the last positional argument.
+
+#### Example with mixed params
+
+```hbs
+<div {{this.myModifier 1 2 op="add"}}></div>
+```
+would be an example invocation of a function with the following signature
+expressed as TypeScript for clarity:
+```ts
+type Options = { op: 'add' | 'subtract' }
+class A {
+  myModifier = (element: Element, first: number, second: number, options: Options) => {
+    // ...
+  }
+}
+```
+
+#### Example with only positional parameters
+
+```hbs
+<div {{this.myModifier 1 2}}></div>
+```
+Because there are no named arguments passed in, the develope may opt to omit the options argument from the method signature.
+```ts
+class A {
+  myModifier = (element: Element, first: number, second: number) => {
+    // ..
+  }
+}
+```
+however, the options argument is still passed, but is equivalent to `{}`.
+
+
+#### Example default modifier implementation
+
+The type signature for any function-modifier could be:
+
+```ts
+import type { CapturedArguments as Arguments } from '@glimmer/interfaces';
+
+type FnArgs<Args extends Arguments = Arguments> =
+  | [Element, ...Args['positional'], Args['named']]
+  | [Element, ...Args['positional'], {}];
+
+type FunctionModifier<
+  Args extends Arguments = Arguments,
+  ModifierArgs extends FnArgs<Args> = FnArgs<Args>
```

Comment:
I think we can simplify this a fair bit, and we also need to capture the case (not common, but does exist) which *only* takes the element.

```ts
type FnArgs<Args extends Arguments = Arguments> =
   | [Element, ...Args['positional'], Args['named']]
   | [Element, ...Args['positional'], {}]
   | [Element];

type FunctionModifier<ModifierArgs extends FnArgs> =
  ((...args: ModifierArgs) => void | (() => void));
```

[(playground)](https://www.typescriptlang.org/play?#code/JYWwDg9gTgLgBDAnmApnA3nAwgQzDAVyhQBMBBKAcwJBQDsYBnOHZi62h5gXzgDMoEEHADkAAUoAbULSgB6YAxRQ+OAMYpGIgNwAoXUlRwAYnXaMAPObgoAHjHok2VGvSZwAvHHauuAPk9dOGCAHzgAbQBRSRROGAAaOAA6FPNwkUhGYBhgCDocSREAXUS0kXzaEmKioNCI6Ni3RJSksszs3PzCkoxuGuC4MKiYuKK9A2Q0YwI6NRy8gFkIEmA+YGULJZW15Ws7BzonEzMqRgCPWoAKS5acU4AuOC3V9ahzAEpPAIA3CGASQZwa6fDw-P4kd7vcZqPKMeCMAAWEAIkhIAGU1IJJJJHtNZvM6M8dlALOEABIAFQWABkGnFEpgAO4I+iPABGEAgMRwdDgfXOQJQkgZcGZrLgiORqIxWMkfPej1+-y+GFqqyBkpR6MxXMkn3QtWCQqSjB12IpEDAnjgAAY9MFuLpHboYXQ4XAYVzcTM5p0ia9SZSaQAFO44ShQPAIuluIoCy5CkEBA1GyQms2SC1WrwARht+b0zt0JBQakkdzQfB9BIQOAA1pp-cpLiBli9lN78X628TA1TqaHIxGozGGHGFXAlSQ9DB642e69Lp69TO54wm1BLprpRmoUA)

Reply:
ooo yes. and with autotracking in modifiers, element-only _is_ common!